src/http-util/src/lib.rs

```diff
@@ -210,3 +212,102 @@ pub async fn handle_tracing() -> impl IntoResponse {
+where
+    I: IntoIterator<Item = &'a HeaderValue>,
+{
+    let allowed = allowed.into_iter().cloned().collect::<Vec<HeaderValue>>();
+    if allowed.iter().any(|o| o.as_bytes() == b"*") {
+        AllowOrigin::any()
+    } else {
+        AllowOrigin::predicate(move |origin: &HeaderValue, _request_parts: _| {
+            for val in &allowed {
+                if (val.as_bytes().starts_with(b"*.")
+                    && origin.as_bytes().ends_with(&val.as_bytes()[1..]))
+                    || origin == val
+                {
+                    return true;
+                }
+            }
+            false
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use http::{header, Method, Request, Response};
+    use hyper::Body;
+    use std::convert::Infallible;
+    use tower::{Service, ServiceBuilder, ServiceExt};
+    use tower_http::cors::CorsLayer;
+
+    #[allow(clippy::unused_async)]
+    async fn handle(_request: Request<Body>) -> Result<Response<Body>, Infallible> {
+        Ok(Response::new(Body::empty()))
+    }
+
+    #[tokio::test]
+    async fn test_cors() -> Result<(), Box<dyn std::error::Error>> {
+        let allowed = vec![
+            HeaderValue::from_str("https://example.com").unwrap(),
+            HeaderValue::from_str("*.example.com").unwrap(),
+        ];
+        let cors = CorsLayer::new()
+            .allow_methods([Method::GET])
+            .allow_origin(build_cors_allowed_origin(allowed.iter()));
+
+        let mut service = ServiceBuilder::new().layer(cors).service_fn(handle);
+        let request = Request::builder()
+            .header(header::ORIGIN, "https://example.com")
+            .body(Body::empty())
+            .unwrap();
+
+        let response = service.ready().await?.call(request).await?;
+
+        assert_eq!(
+            response
+                .headers()
+                .get(header::ACCESS_CONTROL_ALLOW_ORIGIN)
+                .unwrap(),
+            "https://example.com",
+        );
+
+        let request2 = Request::builder()
+            .header(header::ORIGIN, "https://sample.example.com")
+            .body(Body::empty())
+            .unwrap();
+
+        let response2 = service.ready().await?.call(request2).await?;
+
+        assert_eq!(
+            response2
+                .headers()
+                .get(header::ACCESS_CONTROL_ALLOW_ORIGIN)
+                .unwrap(),
+            "https://sample.example.com",
+        );
+
+        let request3 = Request::builder()
+            .method("OPTIONS")
+            .header(header::ORIGIN, "https://materialize.com")
+            .header(header::ACCESS_CONTROL_REQUEST_METHOD, "PUT")
+            .body(Body::empty())
+            .unwrap();
+
+        let response3 = service.ready().await?.call(request3).await?;
+
+        assert!(response3
+            .headers()
+            .get(header::ACCESS_CONTROL_ALLOW_ORIGIN)
+            .is_none());
+
+        Ok(())
+    }
+}
```

Comment:
Nice, thanks for figuring out all the tower related gunk!

This would have been a good moment to use a [table driven test](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests):

```rust
let test_cases = [
    TestCase {
        allowed_origins: vec!["https://example.org"],
        valid_origins: vec!["https://example.org"],
        invalid_origins: vec!["https://wrong.com"],
    },
    TestCase {
        allowed_origins: vec!["https://*.example.org"],
        valid_origins: vec![
            "https://foo.example.org",
            "https://bar.example.org",
            "https://baz.bar.foo.example.org",
        ],
        invalid_origins: vec!["https://example.org", "https://wrong.com"],
    },
    TestCase {
        allowed_origins: vec!["https://*.example.org", "https://wrong.com"],
        valid_origins: vec![
            "https://foo.example.org",
            "https://bar.example.org",
            "https://baz.bar.foo.example.org",
            "https://wrong.com",
        ],
        invalid_origins: vec!["https://example.org"],
    },
    TestCase {
        allowed_origins: vec!["*"],
        valid_origins: vec!["literally", "https://anything.com"],
        invalid_origins: vec![],
    },
    TestCase {
        allowed_origins: vec!["*", "https://iwillbeignored.com"],
        valid_origins: vec!["literally", "https://anything.com"],
        invalid_origins: vec![],
    },
];

for test_case in test_cases {
    let cors = CorsLayer::new()
        .allow_methods([Method::GET])
        .allow_origin(build_cors_allowed_origin(test_case.allowed_origins));
    let mut service = ServiceBuilder::new().layer(cors).service_fn(handle);
    for valid in test_case.valid_origins {
        let response_header = make_request(service, valid);
        assert_eq!(response_header, Some(valid));
    }
    for invalid in test_case.invalid_origins {
        let response_header = make_request(service, valid);
        assert!(response_header.is_none());
    }
}
```

Makes the cost of adding a new test case basically zero, which means more test coverage.

Don't worry about doing that now (better to unblock Robin ASAP), but I may take a stab at it when I get a free minute post merge.

Reply:
Cool, thank you for the tip! I'll go ahead and merge to unblock Robin and out of interest may take a look myself if you don't get to it first :smile: